llvm/lib/SYCLLowerIR/SYCLConditionalCallOnDevice.cpp

```diff
@@ -0,0 +1,128 @@
+//===-- SYCLConditionalCallOnDevice.cpp - SYCLConditionalCallOnDevice Pass
+//--===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Pass performs transformations on functions which represent the conditional
+// call to application's callable object. The conditional call is based on the
+// SYCL device's aspects or architecture passed to the functions.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/SYCLConditionalCallOnDevice.h"
+
+#include "llvm/IR/Function.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Intrinsics.h"
+
+using namespace llvm;
+
+cl::opt<std::string>
+    UniquePrefixOpt("sycl-conditional-call-on-device-unique-prefix",
+                    cl::Optional, cl::Hidden,
+                    cl::desc("Set unique prefix for a translation unit, "
+                             "required for funtions with external linkage"),
+                    cl::init(""));
+
+PreservedAnalyses
+SYCLConditionalCallOnDevicePass::run(Module &M, ModuleAnalysisManager &) {
+  // find call_if_on_device_conditionally function
+  SmallVector<Function *, 4> FCallers;
+  for (Function &F : M.functions()) {
+    if (F.isDeclaration())
+      continue;
+
+    if (CallingConv::SPIR_KERNEL == F.getCallingConv())
+      continue;
+
+    if (F.hasFnAttribute("sycl-call-if-on-device-conditionally") &&
+        ("true" == F.getFnAttribute("sycl-call-if-on-device-conditionally")
+                       .getValueAsString()))
```

Comment:
Do we really need a value here?  I suppose that the attribute is unique enough so that we can simply check for its presence and disregard the value to save on string comparisons.